src/Diagnostics/FxCop/Desktop.Analyzers/Core/Security/DoNotCatchCorruptedStateExceptionsAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+                                    codeBlockStartContext.RegisterCodeBlockEndAction(analyzer.AnalyzeCodeBlockEnd);
+                                }
+                            }
+                        }); 
+                    }
+                });
+        }
+
+        protected abstract class Analyzer
+        {
+            private readonly ISymbol _owningSymbol;
+            private readonly SyntaxNode _codeBlock;
+            private readonly Dictionary<TCatchClauseSyntax, ISymbol> _catchAllCatchClauses;
+
+            public abstract TLanguageKindEnum CatchClauseKind { get; }
+            public abstract TLanguageKindEnum ThrowStatementKind { get; }
+            protected CompilationSecurityTypes TypesOfInterest { get; private set; }
+            protected abstract ISymbol GetExceptionTypeSymbolFromCatchClause(TCatchClauseSyntax catchNode, SemanticModel model);
+            protected abstract bool IsThrowStatementWithNoArgument(TThrowStatementSyntax throwNode);
+            protected abstract bool IsCatchClause(SyntaxNode node);
+            protected abstract bool IslambdaExpression(SyntaxNode node);
+
+            protected Analyzer(CompilationSecurityTypes compilationTypes, ISymbol owningSymbol, SyntaxNode codeBlock)
+            {
+                _owningSymbol = owningSymbol;
+                _codeBlock = codeBlock;
+                _catchAllCatchClauses = new Dictionary<TCatchClauseSyntax, ISymbol>();
+                TypesOfInterest = compilationTypes;
+            }
+
+            public void AnalyzeCatchClause(SyntaxNodeAnalysisContext context)
+            {
+                var catchNode = (TCatchClauseSyntax)context.Node;
+                ISymbol exceptionTypeSymbol = GetExceptionTypeSymbolFromCatchClause(catchNode, context.SemanticModel);
+
+                if (IsCatchTypeTooGeneral(exceptionTypeSymbol))
+                {
+                    var parentNode = catchNode.Parent;
+                    while (parentNode != _codeBlock)
+                    {
+                        // for now there doesn't seem to have any way to annotate lambdas with attributes
+                        if (IslambdaExpression(parentNode))
+                        {
+                            return;
+                        }
+                        parentNode = parentNode.Parent;
+                    }
+                    _catchAllCatchClauses[catchNode] = exceptionTypeSymbol;
+                }
+            }
+
+            public void AnalyzeThrowStatement(SyntaxNodeAnalysisContext context)
+            {
+                var throwNode = (TThrowStatementSyntax)context.Node;
+
+                // throwNode is a throw statement with no argument, which is not allowed outside of a catch clause
+                if (IsThrowStatementWithNoArgument(throwNode))
+                {
+                    var enlcosingCatchClause = (TCatchClauseSyntax)throwNode.Ancestors().First(IsCatchClause);
```

Comment:
@genlu Yes you are right. You should not rely on the order in which the nodes get analyzed, so the only option for you in this scenario is to explicitly look for the enclosing catch clause in the analyzer. I'd still recommend doing it in the analyzer itself for such special cases, and avoid generic helpers so other analyzers inadvertently avoid using them.

Your current implementation looks fine to me.